software/bsg_manycore_lib/bsg_tile_group_barrier.hpp

```diff
@@ -0,0 +1,266 @@
+        return;
+    };
+};
+ 
+
+
+
+
+template <int BARRIER_X_DIM, int BARRIER_Y_DIM>
+class bsg_barrier {
+public:
+    bsg_row_barrier<BARRIER_X_DIM> r_barrier;
+    bsg_col_barrier<BARRIER_Y_DIM> c_barrier;
+
+
+    bsg_barrier () {}
+
+    //  The main sync funciton
+    void sync() {
+        // If barrier dimensions is 1x1, i.e. only a single tile is 
+        // participating, there is nothing to be done
+        if (BARRIER_Y_DIM == 1 && BARRIER_X_DIM == 1)
+                return;
+
+        int center_x_cord = (this->r_barrier._x_cord_start + this->r_barrier._x_cord_end) / 2;
+        int center_y_cord = (this->c_barrier._y_cord_start + this->c_barrier._y_cord_end) / 2;
+
+        #ifdef BSG_BARRIER_DEBUG
+                if( bsg_x == center_x_cord && bsg_y == center_y_cord ){
+                        bsg_print_time();
+                }
+        #endif
+
+        // Send sync signals to the center tile of each row 
+        r_barrier.sync(center_x_cord);
+
+        // Wait on sync signals from all tiles in the row to be received
+        // Send sync signals to the center tile of the center col
+        // Only performed by tiles in the center column
+        if( bsg_x == center_x_cord) {
+                this->r_barrier.wait_on_sync();
+                this->c_barrier.sync(center_x_cord, center_y_cord);
+        }
+
+        // Send alert to all tiles of the col
+        // Wait on sync signals from all tiles in the center column to be received
+        // Send alert signals from the center tile of the center column to all
+        // tiles in the cetner column
+        // Reset the local done list of the bsg_col_barrier class
+        // Only performed by the center tile of the center column
+        if( bsg_x == center_x_cord && bsg_y == center_y_cord) { 
+                this->c_barrier.wait_on_sync();
+                this->c_barrier.reset();
+                this->c_barrier.alert();
+        }
+
+        // Wait on alert signals from center tile of the center column to be received
+        // Send alert signals from center tile of the row to all tiles in the row
+        // Reset the local done list of the bsg_row_barrier class
+        // Performed by all tiles in the center column 
+        if( bsg_x == center_x_cord) {
+                this->c_barrier.wait_on_alert();
+                this->r_barrier.reset();
+                this->r_barrier.alert();
+        }
+
+        // Wait on alert signal from the center tile in each row 
+        // Once the alert signal is received, all tiles are syncrhonized
+        // and can carry on
+        this->r_barrier.wait_on_alert();
+
+        #ifdef BSG_BARRIER_DEBUG
+                if( bsg_x == center_x_cord && bsg_y == center_y_cord ){
+                        bsg_print_time();
+                }
+        #endif
+        return;
+    };
+};
+
+
+// Helper funcitons.
+// check if the char array are all non-zeros
+inline void poll_range( int range, volatile const unsigned int *p){
```

Comment:
is there a reason why we are using int array, when the comment says it's char array? I think we are storing either zero or one, so the char array is okay?

Reply:
Ahh no actually, that's a good point! I changed them to char*, thanks!